book/animations.md

```diff
@@ -370,235 +421,287 @@ browsers.
-update them.
+Importantly, on the next frame, the `SaveLayer` changes but the
+`DrawText`s don't, so on the next frame all we need to do is rerun the
+`SaveLayer`:
 
-``` {.python}
-class BlockLayout:
-	def layout(self):
-		# ...
-        self.width = style_length(
-            self.node, "width", self.parent.width)
-		# ...
-        self.height = style_length(
-            self.node, "height",
-            sum([line.height for line in self.children]))
-```
-
-Here is a simple animation of `width`. As the width of the `div` animates from
-`400px` to `100px`, its height will automatically increase to contain the text
-as it flows into multiple lines.^[And if automatic increase was not desired,
-`height` could be specified to a fixed value. But that would of course cause
-overflow, which needs to be dealt with in one way or another.] Notice how the
-text flows during the animation. It makes sense when resizing, but is otherwise
-confusing and jarring to look at, not to mention hard to read.
-
-``` {.html file=example-width-html}
-<div style="background-color:lightblue;width:100px">
-	This is a test line of text for a width animation.
+    SaveLayer(alpha=0.119866666667)
+      DrawCompositedLayer()
+
+In other words, the idea behind compositing is to split the display
+list into two pieces: a set of composited layers, which are rastered
+during the browser's raster phase and then cached, and a *draw display
+list*, which is drawn during the browser's draw phase and which uses
+the composited layers.
+
+Compositing helps when different frames of the animation have the same
+composited layers, which can then be reused, and only change the draw
+display list. That's the case here, because the only difference
+between frames is the `SaveLayer`, which is in the draw display list.
+Now, a browser can choose what composited layers to create however it
+wants. Typically effects like opacity are very fast to execute on a
+GPU, but *paint commands* that draw shapes---in our browser,
+`DrawText`, `DrawRect`, `DrawRRect`, and `DrawLine`---can be slower.
+Since it's the effects that are typically animated, this means
+browsers usually leave effects in the draw display list and move paint
+commands into composited layers. Of course, in a real browser,
+hardware capabilities, GPU memory, and application data all plays into
+this, but the basic idea of compositing is the same no matter what
+goes where.
+
+Some animations can't be composited because they affect more than just
+the display tree. For example, imagine we animate the `width` of the
+`div` above, instead of animating its opacity. In that case, different
+frames have different *layout trees*, not just display trees. That
+totally changes the coordinates for the `DrawText` calls, and we
+wouldn't necessarily be able to reuse the composited layer. Such
+animations are called *layout-inducing* and speeding them up requires
+[different techniques](reflow.md).[^not-advisable]
```

Comment:
An example  that works in real browsers is here (link to example13-width-transition.html).

Add an exercise to implement `width` animations such that that example works.

Reply:
Great call.

Reply:
Is this done? Don't see it yet.

Reply:
Done now. I got into a nasty git situation.